src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp

```diff
@@ -0,0 +1,259 @@
+                   MultiFluidBase::FluidProp::ViewType const totalDensity )
+      : MultiFluidBase::KernelWrapper( std::move( componentMolarWeight ),
+                                       useMass,
+                                       std::move( phaseFraction ),
+                                       std::move( phaseDensity ),
+                                       std::move( phaseMassDensity ),
+                                       std::move( phaseViscosity ),
+                                       std::move( phaseEnthalpy ),
+                                       std::move( phaseInternalEnergy ),
+                                       std::move( phaseCompFraction ),
+                                       std::move( totalDensity )),
+      m_densities( std::move( densities ) ),
+      m_viscosities( std::move( viscosities ) )
+    {}
+
+  };
+
+  /**
+   * @brief Create a KernelWrapper for this fluid model
+   * @return A KernelWrapper instance for this fluid model
+   */
+  KernelWrapper createKernelWrapper() const;
+
+  /**
+   * @brief Initialize the fluid model with specified number of cells and quadrature points
+   * @param[in] numCells Number of cells/elements
+   * @param[in] numPoints Number of quadrature points per cell
+   */
+  void initialize( localIndex const numCells, localIndex const numPoints )
+  {
+    resizeFields( numCells, numPoints );
+  }
+
+protected:
+  virtual void postInputInitialization() override;
+
+private:
+  array1d< real64 > m_densities;
+  array1d< real64 > m_viscosities;
+};
+
+
+GEOS_HOST_DEVICE
+GEOS_FORCE_INLINE
+void InvariantImmiscibleFluid::KernelWrapper::compute( real64 const pressure,
+                                                       real64 const temperature,
+                                                       arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & composition,
+                                                       PhaseProp::SliceType const phaseFraction,
+                                                       PhaseProp::SliceType const phaseDensity,
+                                                       PhaseProp::SliceType const phaseMassDensity,
+                                                       PhaseProp::SliceType const phaseViscosity,
+                                                       PhaseProp::SliceType const phaseEnthalpy,
+                                                       PhaseProp::SliceType const phaseInternalEnergy,
+                                                       PhaseComp::SliceType const phaseCompFraction,
+                                                       FluidProp::SliceType const totalDensity ) const
+{
+  // Note: pressure and temperature are marked as unused as this model uses constant properties
+  GEOS_UNUSED_VAR( pressure, temperature );
+
+  using Deriv = constitutive::multifluid::DerivativeOffset;
+
+  integer nPhase = phaseDensity.value.size();
```

Comment:
```suggestion
  integer const nPhase = numPhases();
```
Better to rely on this object's properties than caller arguments.